In [ ]:
pip install ktrain

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 25.3/25.3 MB 78.9 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 981.5/981.5 kB 41.4 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.8/468.8 kB 35.5 MB/s eta 0:00:00
  Created wheel for ktrain: filename=ktrain-0.41.4-py3-none-any.whl size=25316932 sha256=b86147dcf76a2c5ab9cb8b5d07f2396540997b3f3f14fc89d826349c42242515
  Stored in directory: /root/.cache/pip/wheels/fa/6a/9c/8a873b38bbd8bc207d33c64726bd18f7ef85f8e70dc3ac2e4b
  Created wheel for keras_bert: filename=ke

In [ ]:
pip install tensorflow==2.11.0

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 588.3/588.3 MB 2.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 75.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 67.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.0/6.0 MB 109.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 439.2/439.2 kB 36.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.9/4.9 MB 117.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 781.3/781.3 kB 54.2 MB/s eta 0:00:00
  Attempting uninstall: tensorboard-data-server
    Found existing installation: tensorboard-data-server 0.7.2
    Uninstalling tensorboard-data-server-0.7.2:
      Successfully uninstalled tensorboard-data-server-0.7.2
  Attempting uninstall: protobuf
    Found existing installation: protobuf 4.25.5
    Uninstalling protobuf-4.25.5:
      Successfully uninstalled protobuf-4.25.5
  Attempting uninstall: keras
    Found existing installation: ke

In [ ]:
import sys
import os
import tensorflow as tf
import re
import string
import nltk as nltk
from nltk.corpus import stopwords
from nltk.stem.wordnet import WordNetLemmatizer
import ktrain
from gensim.models.coherencemodel import CoherenceModel
from gensim.corpora import Dictionary

In [ ]:
import pandas as pd
import re
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.stem import WordNetLemmatizer
from sklearn.feature_extraction.text import CountVectorizer
import itertools
from itertools import combinations
import numpy as np
import matplotlib.pyplot as plt
import random
import spacy

from spacy.language import Language
from collections import Counter

In [ ]:
import gensim
from gensim import corpora
from gensim.models.ldamulticore import LdaMulticore
from gensim.models.coherencemodel import CoherenceModel
from pprint import pprint

In [ ]:
nltk.download('punkt')
nltk.download('stopwords')
nltk.download('wordnet')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package wordnet to /root/nltk_data...


True

In [ ]:
pd.set_option('display.max_rows', 100)
pd.set_option('display.max_columns', None)
pd.set_option('display.max_colwidth', 500)

import warnings
warnings.simplefilter('ignore')

In [ ]:
if torch.cuda.is_available():
    device = "cuda"
else:
    device = "cpu"

print(f"Setting device to '{device}'")

In [ ]:
%%time

from google.colab import drive
import json

# Mount Google Drive
drive.mount('/content/drive')

Mounted at /content/drive
CPU times: user 629 ms, sys: 127 ms, total: 756 ms
Wall time: 15.2 s


In [ ]:
# Define the file path
file_path = '/content/drive/My Drive/Colab Notebooks/NLP_Ignas/nlp_finalproject/df_cleaned_final.csv'

In [ ]:
%%time

# Load the saved CSV file back into notebook as df_cleaned_final
df_cleaned_final = pd.read_csv(file_path)

CPU times: user 11.6 s, sys: 1.59 s, total: 13.1 s
Wall time: 22.7 s


In [ ]:
df_cleaned_final.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 153749 entries, 0 to 153748
Data columns (total 7 columns):
 #   Column       Non-Null Count   Dtype 
---  ------       --------------   ----- 
 0   article_id   153749 non-null  int64 
 1   url          153749 non-null  object
 2   date         153749 non-null  object
 3   language     153749 non-null  object
 4   title        153749 non-null  object
 5   text         153749 non-null  object
 6   token_count  153749 non-null  int64 
dtypes: int64(2), object(5)
memory usage: 8.2+ MB


In [ ]:
df_cleaned_final.head()

article_id  \
0           1   
1           2   
2           3   
3           4   
4           5   

                                                                                                                url  \
0           http://spaceref.com/astronomy/observation-simulation-and-ai-join-forces-to-reveal-a-clear-universe.html   
1                                                             http://www.agoravox.it/Covid-19-un-messaggio-dai.html   
2                                                                                  http://www.dataweek.co.za/21690r   
3                              http://www.envirolink.org/2021/03/11/hundreds-of-sewage-leaks-detected-thanks-to-ai/   
4  http://www.millenniumpost.in/big-stories/ai-express-aircraft-skids-off-runaway-in-kozhikode-rescue-ops-on-414926   

         date language  \
0  2021-07-05       en   
1  2020-03-13       en   
2  2024-04-05       en   
3  2021-03-11       en   
4  2020-08-07       en   

                                                                                    title  \
0         Observation Simulation And AI Join Forces To Reveal A Clear Universe - SpaceRef   
1  Covid-19 un messaggio dai ricercatori italiani ai colleghi stranieri - AgoraVox Italia   
2                               Flash for AI - 28 March 2024 - EBV Electrolink - Dataweek   
3                     Hundreds of sewage leaks detected thanks to AI - EnviroLink Network   
4                               Two dead as AI Express flight skids off Kozhikode airport   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  text  \
0  Observation Simulation And AI Join Forces To Reveal A Clear Universe - SpaceRef Home NASA Watch SpaceRef Business Astrobiology Web Advertising Add an Event Sign up for our Daily Newsletter International Space Station NASA Hack Space Calendar Missions Space Weather Observation Simulation And AI Join Forces To Reveal A Clear Universe Press Release - Source NATIONAL INSTITUTES OF NATURAL SCIENCES Posted July 4 2021 1000 PM View Comments Using AI driven data analysis to peel back the noise and f...   
1  Covid-19 un messaggio dai ricercatori italiani ai colleghi stranieri - AgoraVox Italia Accedi Partecipa ad AgoraVox Iscriviti e proponi un articolo Home page Attualit Ambiente Cronaca Cronaca Locale Cultura Economia Europa Media Istruzione Mondo Politica Salute Religione Societ Scienza e Tecnologia Tribuna Libera Di' la tua Tempo Libero Gossip Redazionali Concorsi Cinema FameTulipani Incredibile ma vero La vignetta del giorno Moda e tendenze Pubblicazioni AgoraVox Recensioni Satira Musica e ...   
2  Flash for AI - 28 March 2024 - EBV Electrolink - Dataweek Home About us Back issues E-book PDF Subscribe Advertise EMP Handbook Categories Editor's Choice Multimedia Videos AI ML Analogue Mixed Signal LSI Circuit System Protection ComputerEmbedded Technology Design Automation DSP Micros Memory Edge Computing IIoT Electronics Technology Enclosures Racks Cabinets Panel Products Events Interconnection Manufacturing Production Technology Hardware Services News Opto-Electronics Passive Components...   
3  Hundreds of sewage leaks detected thanks to AI - EnviroLink Network Home EnviroLink Forum About EnviroLink Thursday 11 March 2021 You are here Home 2021 March 11 Hundreds of sewage leaks detected thanks to AI Hundreds of sewage leaks detected thanks to AI Posted on March 11 2021 by EnviroLink Editor 0 Comments From BBC Shareclose Share pageCopy linkAbout sharing image copyrightRivers Trust Hundreds of previously unreported releases of raw

Data Cleaning

In [ ]:
df = df_cleaned_final.copy()

In [ ]:
lemmatizer = WordNetLemmatizer()
stop_words = set(stopwords.words('english'))

def clean_text(text):
    """
    Cleans text by:
    - Removing special characters and numbers.
    - Converting to lowercase.
    - Removing stopwords.
    - Lemmatizing the words to their base form.
    Returns a single string representing the cleaned text.
    """

    # Remove special characters and numbers
    text = re.sub(r'[^A-Za-z\s]', '', text)

    # Convert to lowercase
    text = text.lower()

    # Remove stopwords and lemmatize
    words = text.split()
    cleaned_words = [lemmatizer.lemmatize(word) for word in words if word not in stop_words]

    # Return the cleaned text as a single string
    return ' '.join(cleaned_words)

In [ ]:
# Apply to 'text' and 'title' column
df['text_cleaned'] = df['text'].apply(clean_text)
df['title_cleaned'] = df['title'].apply(clean_text)

In [ ]:
# Combine title and text to create a new column called "title_text_cleaned"
df['title_text_cleaned'] = df['text_cleaned'] + " " + df['title_cleaned']

In [ ]:
# Turn the column 'title_text_cleaned' to a list for further analysis
# A list of documents (strings). The topic model will be trained on these documents.
texts = df['title_text_cleaned'].tolist()

### Start Topic Modeling with 12 topics

A topic is not a single word but rather a distribution of words.

For example:

Topic 1 might consist of words like {"economy": 0.2, "trade": 0.1, "inflation": 0.05, ...}.

Topic 2 might include words like {"technology": 0.3, "innovation": 0.1, "AI": 0.05, ...}.

When a document is analyzed, the topic model determines the proportion of the document that belongs to each topic.

In [ ]:
%%time
tm = ktrain.text.get_topic_model(
    texts=texts,
    n_topics=12,
    n_features=10000, # Limits the vocabulary size to the top 10,000 most frequent terms in the dataset.
    min_df=0.001, # Words that appear in less than 0.1% of the documents are ignored.
    max_df=0.5, # Words that appear in more than 50% of the documents are ignored.
    stop_words='english', # Common English stop words (like "the," "and," "is") are removed before training to focus on meaningful terms.
    model_type='lda',
    lda_max_iter=5, # The model will iterate up to 5 times to converge on a solution for topic distributions.
    verbose=1) # Enables logging to show progress and information during training. It helps monitor the training process.

lang: en
preprocessing texts...
fitting model...
iteration: 1 of max_iter: 5
iteration: 2 of max_iter: 5
iteration: 3 of max_iter: 5
iteration: 4 of max_iter: 5
iteration: 5 of max_iter: 5
done.
CPU times: user 13min 57s, sys: 4min 3s, total: 18min
Wall time: 13min 20s


In [ ]:
tm.print_topics()

topic 0 | music public program radio schedule event art community search npr
topic 1 | google user best feature video apple image tech microsoft search
topic 2 | market global analysis growth research forecast key trend player size
topic 3 | risk government security job tech law potential generative tool work
topic 4 | release newswires south press north ein distribution presswire island country
topic 5 | share india price day latest bank daily read market subscribe
topic 6 | customer cloud solution platform model enterprise generative application experience digital
topic 7 | product entertainment consumer release resource general health public financial overviewview
topic 8 | human science like learning machine people used model using health
topic 9 | gray group release press solution prnewswire alert statement platform patient
topic 10 | stock market nasdaq investor symbol quote investment fund insurance analyst
topic 11 | ago video hour weather sport story day local public search


In [ ]:
# For each document in the dataset (texts), it calculates the topic probabilities using the trained model.
# If the most probable topic for a document has a probability equal to or greater than 0.25, that topic is assigned to the document.
# Purpose: The build method assigns topics to the input texts (documents) based on the trained topic model (tm).
# For each document in the dataset (texts), it calculates the topic probabilities using the trained model.
# Assigns topics to documents based on the threshold parameter.

%%time
tm.build(texts, threshold=0.25)

done.
CPU times: user 3min 36s, sys: 40.3 s, total: 4min 17s
Wall time: 3min 30s


In [ ]:
tm.print_topics(show_counts=True)

topic:9 | count:26215 | gray group release press solution prnewswire alert statement platform patient
topic:8 | count:19444 | human science like learning machine people used model using health
topic:6 | count:19118 | customer cloud solution platform model enterprise generative application experience digital
topic:1 | count:17554 | google user best feature video apple image tech microsoft search
topic:11 | count:17230 | ago video hour weather sport story day local public search
topic:3 | count:13980 | risk government security job tech law potential generative tool work
topic:5 | count:10742 | share india price day latest bank daily read market subscribe
topic:2 | count:7697 | market global analysis growth research forecast key trend player size
topic:10 | count:6800 | stock market nasdaq investor symbol quote investment fund insurance analyst
topic:0 | count:5472 | music public program radio schedule event art community search npr
topic:4 | count:4393 | release newswires south press nor

Visualize the topic distributions for documents after training a topic model (tm) in ktrain

In [ ]:
# Retrieves the topic distributions for all documents in the dataset
# A matrix where each row corresponds to a document and each column corresponds to a topic. The values are the probabilities of each topic for the document.
doc_topics=tm.get_doctopics()

In [ ]:
# Visualize the documents in a 2D space based on their topic distributions.
# Reduces the dimensionality of the doc_topics matrix (e.g., using t-SNE or UMAP) to project documents into 2D.
# Each document is represented as a point in this 2D space, with colors or labels indicating the dominant topic for each document.
# Clusters of Documents: Documents with similar topic distributions will be closer to each other.
# Dominant Topics: Points may be color-coded by the most probable topic, helping you identify clusters for each topic.
tm.visualize_documents(doc_topics=tm.get_doctopics())

Output hidden; open in https://colab.research.google.com to view.

### Assigns topics to documents based on the topic probabilities predicted by the topic model (tm)

In [ ]:
# Generates the topic probability distribution for each document in texts using the trained topic model (tm).
# A 2D array (topic_probs) where each row corresponds to a document and each column corresponds to a topic.
# Each value is the probability of the document belonging to that topic.
topic_probs = tm.predict(texts)

# Specifies the minimum probability a topic must have to be assigned to a document.
# Only topics with probabilities ≥ 0.25 will be considered.
threshold = 0.25

# An empty list to store the assigned topic for each document.
topic_assignments = []

# For each document, select the most probable topic and check if it exceeds the threshold.
# Loops through each document's topic probabilities and assigns a topic based on the threshold.
for probs in topic_probs:
    top_topic = np.argmax(probs)  # Select the most probable topic
    if probs[top_topic] >= threshold:  # If the probability of that topic is above the threshold
        topic_assignments.append(top_topic)  # Assign the topic
    else:
        topic_assignments.append(None)  # Assign None if below threshold

# Add a new column to the dataframe containing the predicted topic indices
df['topic'] = topic_assignments

In [ ]:
# Replace NaN values in the 'topic' column with 'other'
df['topic'] = df['topic'].fillna('other')

# Convert all values in the 'topic' column to strings
df['topic'] = df['topic'].astype(str)

In [ ]:
# The number of rows (articles) for each topic
df['topic'].value_counts()

topic
9.0      26215
8.0      19444
6.0      19118
1.0      17554
11.0     17230
3.0      13980
5.0      10742
2.0       7697
10.0      6800
0.0       5472
4.0       4393
7.0       3717
other     1387
Name: count, dtype: int64

Art: Topic 0

MarketResearch: Topic 2, Topic 5

Media: Topic 4

PolicyandLegal: Topic 3

Finance: Topic 10

Topics that are related to tech: Topic 6, Topic 8, and Topic 1

Keywords in the topic seem to have no relevance (other): Topic 7, Topic 9, and Topic 11

In [ ]:
df.head(1)

article_id  \
0           1   

                                                                                                       url  \
0  http://spaceref.com/astronomy/observation-simulation-and-ai-join-forces-to-reveal-a-clear-universe.html   

         date language  \
0  2021-07-05       en   

                                                                             title  \
0  Observation Simulation And AI Join Forces To Reveal A Clear Universe - SpaceRef   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  text  \
0  Observation Simulation And AI Join Forces To Reveal A Clear Universe - SpaceRef Home NASA Watch SpaceRef Business Astrobiology Web Advertising Add an Event Sign up for our Daily Newsletter International Space Station NASA Hack Space Calendar Missions Space Weather Observation Simulation And AI Join Forces To Reveal A Clear Universe Press Release - Source NATIONAL INSTITUTES OF NATURAL SCIENCES Posted July 4 2021 1000 PM View Comments Using AI driven data analysis to peel back the noise and f...   

   token_count  \
0          912   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          text_cleaned  \
0  observation simulation ai join force reveal clear universe spaceref home nasa watch spaceref business astrobiology web advertising add event sign daily newsletter international space station nasa hack space calendar mission space weather observation simulation ai join force reveal clear universe press release source national institute natural science posted july pm view comment using ai driven data analysis peel back noise find actual shape universe credit institute statistical mathematics j...   

                                                         title_cleaned  \
0  observation simulation ai join force reveal clear universe spaceref   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    title_text_cleaned  \
0  observation simulation ai join force reveal clear universe spaceref home nasa watch spaceref business astrobiology web advertising add event sign daily newsletter international space station nasa hack space calendar mission space weather observation simulation ai join force reveal clear universe press release source national institute natural science posted july pm view comment using ai driven data analysis peel back noise find actual shape universe credit institute statistical mathematics j...   

  topic  
0   8.0

### I want to focus on Topics that are related to tech: Topic 6, Topic 8, and Topic 1

In [ ]:
filtered_df = df[df['topic'].isin(['1.0', '6.0', '8.0'])]

In [ ]:
filtered_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 56116 entries, 0 to 153748
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   article_id          56116 non-null  int64 
 1   url                 56116 non-null  object
 2   date                56116 non-null  object
 3   language            56116 non-null  object
 4   title               56116 non-null  object
 5   text                56116 non-null  object
 6   token_count         56116 non-null  int64 
 7   text_cleaned        56116 non-null  object
 8   title_cleaned       56116 non-null  object
 9   title_text_cleaned  56116 non-null  object
 10  topic               56116 non-null  object
dtypes: int64(2), object(9)
memory usage: 5.1+ MB


In [ ]:
texts = filtered_df['title_text_cleaned'].tolist()

In [ ]:
# Try n=10

%%time
tm = ktrain.text.get_topic_model(
    texts=texts,
    n_topics=10,
    n_features=10000,
    min_df=0.001,
    max_df=0.5,
    stop_words='english',
    model_type='lda',
    lda_max_iter=5,
    verbose=1)

lang: en
preprocessing texts...
fitting model...
iteration: 1 of max_iter: 5
iteration: 2 of max_iter: 5
iteration: 3 of max_iter: 5
iteration: 4 of max_iter: 5
iteration: 5 of max_iter: 5
done.
CPU times: user 4min 33s, sys: 440 ms, total: 4min 33s
Wall time: 4min 33s


In [ ]:
tm.print_topics()

topic 0 | best review game icon home tech gaming mobile video smart
topic 1 | health patient machine human algorithm learning risk study medical potential
topic 2 | tool user experience customer video generative marketing image create platform
topic 3 | market industry cooky event said tech startup digital cookie investment
topic 4 | google apple feature microsoft pro window iphone user search phone
topic 5 | science learning machine research researcher university scientist computer follow image
topic 6 | customer solution cloud platform application enterprise gray security group generative
topic 7 | april march september june july december october january february august
topic 8 | said people say human image way thing day think chatgpt
topic 9 | india event product channel wire password email meta digital home


In [ ]:
%%time
tm.build(texts, threshold=0.25)

done.
CPU times: user 1min 13s, sys: 266 ms, total: 1min 13s
Wall time: 1min 13s


In [ ]:
tm.print_topics(show_counts=True)

topic:8 | count:12473 | said people say human image way thing day think chatgpt
topic:6 | count:10652 | customer solution cloud platform application enterprise gray security group generative
topic:4 | count:6585 | google apple feature microsoft pro window iphone user search phone
topic:2 | count:5000 | tool user experience customer video generative marketing image create platform
topic:3 | count:4996 | market industry cooky event said tech startup digital cookie investment
topic:1 | count:4474 | health patient machine human algorithm learning risk study medical potential
topic:5 | count:4081 | science learning machine research researcher university scientist computer follow image
topic:0 | count:2409 | best review game icon home tech gaming mobile video smart
topic:7 | count:2392 | april march september june july december october january february august
topic:9 | count:2194 | india event product channel wire password email meta digital home


In [ ]:
# Display the text of the most relevant document for each topic index from 0 to 9.
for topic_id in range(10):
    # Retrieve documents related to each topic and display the text of the most relevant document.
    print(f"Top document for Topic {topic_id}:")
    print(tm.get_docs(topic_ids=[topic_id], rank=True)[0]['text'][:1000])
    print("\n")

Top document for Topic 0:
un lg uhd tv thinq ai unpub lg ca javascript appears disabled browser please enable javascript ensure get lg site product service cooky site use cooky including cooky third party enhance user experience effectiveness marketing activity cooky performance analytics advertising cooky please see privacy cookie policy information agree cooky select accept select cookie setting see cooky use choose one would like accept cookie setting accept cooky setting absolutely necessary cooky absolutely necessary cooky cooky absolutely essential running website category cooky cannot disabled functional cooky functional cooky cooky used provide convenient function product review product video playback web browsing also enable interaction social medium account social medium social medium connect social channel share lg technology friend example facebook twitter analytics cooky analytics cooky cooky allow website keep improving function website traffic analysis suggest personaliz

In [ ]:
doc_topics=tm.get_doctopics()
tm.visualize_documents(doc_topics=tm.get_doctopics())

Output hidden; open in https://colab.research.google.com to view.

topic:9 | count:2194 | india event product channel wire password email meta digital home - is widely spread

**Distance Matters:**

Documents with similar topic probabilities will appear closer in 2D.

Documents dominated by different topics will appear farther apart.

**Clusters Represent Similarities:**

Documents in the same region are likely to share similar dominant topics.
Overlaps in clusters indicate multi-topic documents.

**Dimensionality Reduction Artifacts:**

While relative positions are preserved, the 2D representation is an approximation and may not perfectly reflect distances in the original topic space.

In [ ]:
# Get the topic probability distribution
topic_probs = tm.predict(texts)

# Set the threshold
threshold = 0.25

# Prepare an empty list for topic assignments
topic_assignments = []

# For each document, select the most probable topic and check if it exceeds the threshold
for probs in topic_probs:
    top_topic = np.argmax(probs)  # Select the most probable topic
    if probs[top_topic] >= threshold:  # If the probability of that topic is above the threshold
        topic_assignments.append(top_topic)  # Assign the topic
    else:
        topic_assignments.append(None)  # Assign None if below threshold


# Add a new column to the dataframe containing the predicted topic indices
filtered_df['topic_2'] = topic_assignments

In [ ]:
# Replace NaN values in the 'topic' column with 'other'
filtered_df['topic_2'] = filtered_df['topic_2'].fillna('other')

# Convert all values in the 'topic' column to strings
filtered_df['topic_2'] = filtered_df['topic_2'].astype(str)

In [ ]:
filtered_df['topic_2'].value_counts()

topic_2
8.0      12473
6.0      10652
4.0       6585
2.0       5000
3.0       4996
1.0       4474
5.0       4081
0.0       2409
7.0       2392
9.0       2194
other      860
Name: count, dtype: int64

EntertainmentTech: Topic 0, Topic 8

HealthcareTech: Topic 1

AIResearch: Topic 5

TechProducts: Topic 4

CloudComputing: Topic 6

AIinContentCreation: Topic 2


Keywords in the topic seem to have no relevance (other): Topic 7, Topic 3, Topic 9

### With above, I will perform topic labeling

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 153749 entries, 0 to 153748
Data columns (total 11 columns):
 #   Column              Non-Null Count   Dtype 
---  ------              --------------   ----- 
 0   article_id          153749 non-null  int64 
 1   url                 153749 non-null  object
 2   date                153749 non-null  object
 3   language            153749 non-null  object
 4   title               153749 non-null  object
 5   text                153749 non-null  object
 6   token_count         153749 non-null  int64 
 7   text_cleaned        153749 non-null  object
 8   title_cleaned       153749 non-null  object
 9   title_text_cleaned  153749 non-null  object
 10  topic               153749 non-null  object
dtypes: int64(2), object(9)
memory usage: 12.9+ MB


In [ ]:
df['topic'].value_counts()

topic
9.0      26215
8.0      19444
6.0      19118
1.0      17554
11.0     17230
3.0      13980
5.0      10742
2.0       7697
10.0      6800
0.0       5472
4.0       4393
7.0       3717
other     1387
Name: count, dtype: int64

In [ ]:
# Define a dictionary to map topic numbers to topic names
topic_mapping = {
    '0.0': 'Art',
    '1.0': 'topic_2',
    '2.0': 'MarketResearch',
    '3.0': 'PolicyandLegal',
    '4.0': 'Media',
    '5.0': 'MarketResearch',
    '6.0': 'topic_2',
    '7.0': 'other',
    '8.0': 'topic_2',
    '9.0': 'other',
    '10.0': 'Finance',
    '11.0': 'other',
    'other': 'other'
}

# Map the topic numbers to topic names in the 'topic' column
df['topic'] = df['topic'].map(topic_mapping)

In [ ]:
filtered_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 56116 entries, 0 to 153748
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   article_id          56116 non-null  int64 
 1   url                 56116 non-null  object
 2   date                56116 non-null  object
 3   language            56116 non-null  object
 4   title               56116 non-null  object
 5   text                56116 non-null  object
 6   token_count         56116 non-null  int64 
 7   text_cleaned        56116 non-null  object
 8   title_cleaned       56116 non-null  object
 9   title_text_cleaned  56116 non-null  object
 10  topic               56116 non-null  object
 11  topic_2             56116 non-null  object
dtypes: int64(2), object(10)
memory usage: 5.6+ MB


In [ ]:
# Define a dictionary to map topic numbers to topic names
topic2_mapping = {
    "0.0": "EntertainmentTech",
    "1.0": "HealthcareTech",
    "2.0": "AIinContentCreation",
    "3.0": "other",
    "4.0": "TechProducts",
    "5.0": "AIResearch",
    "6.0": "CloudComputing",
    "7.0": "other",
    "8.0": "EntertainmentTech",
    "9.0": "other",
    "other": "other"
}

# Map the topic numbers to topic names in the 'topic2' column
filtered_df['topic_2'] = filtered_df['topic_2'].map(topic2_mapping)

In [ ]:
# Merge data
merged_df = df.merge(filtered_df[['article_id', 'topic_2']], on='article_id', how='left')

In [ ]:
# Refine the final topic column
# If the value in topic is 'topic_2', it replaces it with the corresponding value from topic_2.

# merged_df.loc[merged_df['topic'] == 'topic_2', 'topic']: Selects the topic column for rows where the condition is True (i.e., rows with 'topic_2' as their value).
# merged_df.loc[merged_df['topic'] == 'topic_2', 'topic_2']: Fetches the corresponding values from the topic_2 column for the same rows where topic is 'topic_2'.

# The topic column is updated with the corresponding values from the topic_2 column for these specific rows.
merged_df.loc[merged_df['topic'] == 'topic_2', 'topic'] = merged_df.loc[merged_df['topic'] == 'topic_2', 'topic_2']

In [ ]:
# Topic Count
merged_df['topic'].value_counts()

topic
other                  58991
MarketResearch         18439
EntertainmentTech      14882
PolicyandLegal         13980
CloudComputing         10652
Finance                 6800
TechProducts            6585
Art                     5472
AIinContentCreation     5000
HealthcareTech          4474
Media                   4393
AIResearch              4081
Name: count, dtype: int64

In [ ]:
merged_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 153749 entries, 0 to 153748
Data columns (total 12 columns):
 #   Column              Non-Null Count   Dtype 
---  ------              --------------   ----- 
 0   article_id          153749 non-null  int64 
 1   url                 153749 non-null  object
 2   date                153749 non-null  object
 3   language            153749 non-null  object
 4   title               153749 non-null  object
 5   text                153749 non-null  object
 6   token_count         153749 non-null  int64 
 7   text_cleaned        153749 non-null  object
 8   title_cleaned       153749 non-null  object
 9   title_text_cleaned  153749 non-null  object
 10  topic               153749 non-null  object
 11  topic_2             56116 non-null   object
dtypes: int64(2), object(10)
memory usage: 14.1+ MB


In [ ]:
merged_df.drop(columns=['topic_2'], inplace=True)

In [ ]:
merged_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 153749 entries, 0 to 153748
Data columns (total 11 columns):
 #   Column              Non-Null Count   Dtype 
---  ------              --------------   ----- 
 0   article_id          153749 non-null  int64 
 1   url                 153749 non-null  object
 2   date                153749 non-null  object
 3   language            153749 non-null  object
 4   title               153749 non-null  object
 5   text                153749 non-null  object
 6   token_count         153749 non-null  int64 
 7   text_cleaned        153749 non-null  object
 8   title_cleaned       153749 non-null  object
 9   title_text_cleaned  153749 non-null  object
 10  topic               153749 non-null  object
dtypes: int64(2), object(9)
memory usage: 12.9+ MB


In [ ]:
# Define the file path
file_path = '/content/drive/My Drive/Colab Notebooks/NLP_Ignas/nlp_finalproject/topic_modeling.csv'

In [ ]:
# Save the DataFrame as a CSV file
merged_df.to_csv(file_path, index=False)